hibernate-core/src/main/java/org/hibernate/dialect/OracleDialect.java

```diff
@@ -1103,16 +1146,16 @@ public static Replacer datetimeFormat(String format, boolean useFm) {
 				//hour
 				.replace("hh", "HH12")
 				.replace("HH", "HH24")
-				.replace("h", fm + "HH12")
-				.replace("H", fm + "HH24")
+				.replace("h", fm + "HH12" + fm)
+				.replace("H", fm + "HH24" + fm)
 
 				//minute
 				.replace("mm", "MI")
-				.replace("m", fm + "MI")
+				.replace("m", fm + "MI" + fm)
 
 				//second
 				.replace("ss", "SS")
-				.replace("s", fm + "SS")
+				.replace("s", fm + "SS" + fm)
```

Comment:
I don't understand the reason for this change. According to the Oracle documentation, the leading `fm` suppresses trailing blanks as well as leading zeros. Why did you add this?

Also, note that this function is called by other dialects; we should move it somewhere else.

Reply:
According to the Oracle documentation this is a "trigger" to switch the behavior and it can occur multiple times. PostgreSQL seems to reset this trigger when encountering a "literal", just as I would have expected from Oracle, but apparently Oracle keeps this mode, so it needs to be reset, otherwise the rest of the pattern will use the modifier as well.

Reply:
So you're saying you include it to *undo* the effect of `fm`?

But in fact we never wish to turn it off, it seems to me. If it could be one setting at the start of the pattern that would be fine by me. Right?

Reply:
I'm saying that if you just use `fm` once in Oracle, every following pattern will trim leading zeros even if your desired format should retain them. To avoid this, you have to "switch" again by using the fm modifier again.

Reply:
Also see https://docs.oracle.com/cd/B19306_01/server.102/b14200/sql_elements004.htm#SQLRF00216

Reply:
> every following pattern will trim leading zeros even if your desired format should retain them

but in what circumstance do we ever want to retain them. IIRC our `format()` function always has the semantics of no padding, doesn't it?

(now where did I document that??)

Reply:
AFAICT our `format()` uses the same format like `java.time.format.DateTimeFormatter` and we translate that to the DBMS specific format. To get the semantics right, it is necessary to "reset"/"switch" the modifier by using `fm` again.

Reply:
Sorry, you're right and I'm an idiot. The main reason we're doing this is because we _sometimes_ want to eliminate leading zeros. The purpose isn't the stupid trailing spaces in Oracle. I forgot.

Still not certain we want to add the trailing `fm`s for other dialects apart from Oracle though. I think we should add another boolean toggle to the method signature, so this is something we only do on Oracle.

Reply:
FTR, the only documentation I wrote for this feature (so far) is the Javadoc in `Dialect.datetimeFormat()`.

Reply:
Ok, I'll adapt it for Oracle then. It's just that it's not "wrong" to toggle the `fm` mode off by using it again. All DBMS do the right thing. It seems that at least DB2 and PostgreSQL reset the `fm` mode automatically when encountering a literal, but not sure if all DBMS work this way. At the same time, these DBMS also handle a trailing `fm` correctly in the sense that it also toggles the `fm` mode.

Reply:
Whatever you decide to do is fine by me. My original objection to this was wrong.